src/main/java/com/github/princesslana/eriscasper/BotContext.java

```diff
@@ -51,21 +55,20 @@ public Completable doNothing() {
    *
    * <p>(As opposed to a Query)
    */
-  public <I, O> Completable execute(Action<I, O> action) {
-    LOG.debug("Executing {}...", action);
-    return execute(action.getRoute(), action.getData()).toCompletable();
+  public Completable execute(Action action) {
+    return Single.just(actionContext).flatMapCompletable(action);
```

Comment:
Why not just `action.apply(actionContext)`?

Reply:
`io.reactivex.function.Function` produces an `Exception` exception on `Function#apply(X)`. The way to handle that error respectively in a Completable context is by passing it through a `Single`.

Reply:
As an option here, for `Bot`, we extend `Function`, but define it without an exception: https://github.com/princesslana/ErisCasper.java/blob/master/src/main/java/com/github/princesslana/eriscasper/Bot.java#L19

Reply:
I think keeping in the checked exception is good, I think it produces the exception in a clean manner. Especially since we're providing a `Completable` context. We want to cleanly handle all exceptions thrown by `Routes` or the `Gateway` inside of the stream. I think it's good how it is. If it's a problem or hurts efficiency I can change it, but I think it's the cleanest thing to do.

Reply:
Especially for making it an exposed API, so the user can throw exceptions inside the stream without being worried about them being handled poorly.

Reply:
My thinking is this:
* The only reason we use reactivex `Function` instead of Java `Function` is because it works nicer with the reactivex streams.
* In general `throws Exception` is bad, and along with  https://rongi.github.io/kotlin-blog/rxjava/rx/2017/08/01/error-handling-in-rxjava.html, I don't think we want our Actions throwing an exception.

Reply:
That's a fair point. I'll change it.